importer/helpers/dagbuilder.go

```diff
@@ -202,12 +259,63 @@ func (db *DagBuilderHelper) GetNextDataNode() (*UnixfsNode, error) {
 	return db.NewLeaf(data)
 }
 
-// SetPosInfo sets the offset information of a node using the fullpath and stat
-// from the DagBuilderHelper.
-func (db *DagBuilderHelper) SetPosInfo(node *UnixfsNode, offset uint64) {
+// NewLeafDataNode is a variation of `GetNextDataNode` that returns
+// an `ipld.Node` instead. It builds the `node` with the data obtained
+// from the Splitter and returns it with the `dataSize` (that will be
+// used to keep track of the DAG file size). The size of the data is
+// computed here because after that it will be hidden by `NewLeafNode`
+// inside a generic `ipld.Node` representation.
+func (db *DagBuilderHelper) NewLeafDataNode() (node ipld.Node, dataSize uint64, err error) {
+	fileData, err := db.Next()
+	if err != nil {
+		return nil, 0, err
+	}
+	dataSize = uint64(len(fileData))
+
+	// Create a new leaf node containing the file chunk data.
+	node, err = db.NewLeafNode(fileData)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Convert this leaf to a `FilestoreNode` if needed.
+	node = db.ProcessFileStore(node, dataSize)
```

Comment:
the downside of doing things this way is that we hold the node in memory for longer than we really need to (technically just need to hash the data, could be done entirely on stack if go let us)

Reply:
I'm not following, *this way* refers to this PR or in general? From what I understand, we normally hold work-in-progress nodes to avoid constantly (un)packing while working on them, reducing computation time at the cost of more memory, but I'm not sure if you're referring to that.

Reply:
Yes, something that I'm seeing throughout the entire code is this mixture of node structures that are used interchangeably for (i) reading/writing and for (ii) building/cache. Ideally, if you are retrieving content (or storing what you received from another peer) the first type is just fine, but if you are building stuff yourself (and that is the very common operation of just adding content with `ipfs add`) then you need an entirely different object, that yes, will look like the first one (and the first one will even be its end product) but actually will behave very differently: it will cache many things while in the process of modifying them, it will have information in a mutable state, it will allow (what the first type would see as) invalid states.

Ideally again, I would expect to have different structures (with a common base or shared methods) to reflect that. And going back to the original raw node comment, yes, someone reading (or writing) the node needs the buffer, but someone building it may only need partial information like the hash.

Sorry, I'm not sure if you're original remark was going in this direction, but it made me think of this situation that I was encountering repeatedly while working on this PR.